File: OrangeJuice.Server.Core/Data/JsonProductDescriptorConverter.cs
﻿using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

using OrangeJuice.Server.Data.Models;

namespace OrangeJuice.Server.Data
{
	public sealed class JsonProductDescriptorConverter : IConverter<string, ProductDescriptor>
	{
		public ProductDescriptor Convert(string value)
		{
			return JObject.Parse(value)
						  .ToObject<ProductDescriptor>();
		}

		public string ConvertBack(ProductDescriptor value)
		{
			return JObject.FromObject(value)
						  .ToString(Formatting.Indented);
		}
	}
}

File: OrangeJuice.Server.Core.FSharp/Data/JsonProductDescriptorConverter.fs
namespace OrangeJuice.Server.FSharp.Data

open Newtonsoft.Json
open Newtonsoft.Json.Linq

open OrangeJuice.Server
open OrangeJuice.Server.Data

type JsonProductDescriptorConverter() =
    interface IConverter<string, ProductDescriptor> with
        member this.Convert(value : string) : ProductDescriptor =
            JObject.Parse(value).ToObject<ProductDescriptor>()
        member this.ConvertBack(value : ProductDescriptor) : string =
            JObject.FromObject(value).ToString(Formatting.Indented)
